doc/tryto.1
.TH tryto 1
.SH NAME
tryto - tries to run a command limited by a timeout or number of tries,
can be used to run multilog(8) processors.
.SH SYNOPSIS
.B tryto
[
.B \-pv
]
[
.B \-t
.I seconds
]
[
.B \-n
.I tries
]
.I prog
.SH DESCRIPTION
.B tryto
runs and watches
.IR prog .
If
.I prog
exits with an error code other then 0,
.B tryto
runs
.I prog
again after sleeping one second.
.P
If the number of retries reaches the maximal
number of
.IR tries ,
.B tryto
prints an error message and gives up.
.P
If the timeout
.I seconds
is reached and
.I prog
is still running,
.B tryto
sends a TERM signal to
.IR prog ,
waits 5 seconds for
.I prog
to terminate and then send a KILL signal if
.I prog
still is there, end exits as soon as possible.
.SH OPTIONS
.TP
.B \-v
verbose. print verbose messages to stderr.
.TP
.B \-p
processor. Use this option if you run
.B tryto
as a
.BR multilog (8)
processor (see below).
.TP
.B \-t \fIseconds
timeout. Set the timeout to
.I seconds
seconds. Default is 180.
.TP
.B \-n \fItries
Set the maximal number of tries to
.IR tries .
If
.I prog
has crached,
.B tryto
tries to rewind stdin to the beginning using
.IR lseek (2),
before starting
.I prog
again. Default is 5.
.SH MULTILOG PROCESSOR
If
.B tryto
sees the
.B -p
option,
.B tryto
runs as a
.BR multilog (8)
processor, making use of filedescriptors 4 and 5:
.P
After forking,
.B tryto
moves the filedescriptor 5 to 2, so all error messages from
.B tryto
and
.I prog
will be saved in
.BR multilog 's(8)
.I state
to be processed on the next run of
.BR tryto .
.P
After starting
.IR prog ,
.B tryto
first feeds all data it reads from filedescriptor 4 to
.IR prog 's
standard input, then all data from filedescriptor 0.
.P
If
.I prog
fails by timeout
.I seconds
or maximal number of
.IR tries ,
.B tryto
prints all data from standard input to standard output, an error message
to filedesriptor 5 and exits with 0.
.SH EXIT CODES
If
.B tryto
runs as a
.BR multilog (8)
processor,
.B tryto
always returns 0.
.P
If
.I prog
was run successfully,
.B tryto
returns 0.
.P
If
.I prog
failed by timeout,
.B tryto
returns 10.
.P
If
.I prog
failed by maximal number of
.IR tries ,
.B tryto
returns the last exit code from
.IR prog .
.P
If
.B tryto
itself failes, it exists with 111.
.SH SEE ALSO
multilog(8),
socklog(8),
lseek(2)

http://innominate.org/~pape/socklog/
.SH AUTHOR
G. Pape <pape@innominate.com>


package/CHANGES
  * doc/*.html: add color-scheme.
  * doc/benefits.html, doc/notify.html: use chpst -u instead of setuidgid.
  * doc/configuration.html, doc/notify.html, doc/upgrade.html,
    man/socklog-conf.8: use /service instead of /var/service.
  * doc/index.html: spelling.
  * doc/install.html: how to install as non-root user.
  * doc/examples.html, doc/notify.html: upper and lower case, format.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * socklog-check.c, socklog.c: check length of unix socket to match struct
    sockaddr.
  * pathexec_env.c, socklog-conf.c, socklog.c, tryto.c, uncat.c: whitespace.
  * socklog.c: fix compilation warning -Wpointer-sign.
  * socklog.c, man/socklog.8: add -T to drop timestamps from syslog messages
    (github/leahneukirchen).
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

socklog 2.1.2
Sun, 02 Nov 2025 11:41:11 +0100
  * alloc.c, alloc_re.c, buffer_0.c, byte_chr.c, byte_copy.c, byte_cr.c,
    byte_diff.c, byte_rchr.c, byte_zero.c, error.h, select.h2, wait.h,
    wait_nohang.c, wait_pid.c: sync daemontools library fork with runit
    source.
  * doc/*.html: !DOCTYPE.
  * socklog.c, tryto.c, uncat.c: use parameter name "unused" in sighandler
    functions.
  * hasflock.h1: add #include <unistd.h>.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-socklog-unix.c, pathexec_env.c, select.h1, select.h2,
    sig.c, sig.h, sig_catch.c, socklog-check.c, socklog-check.dist,
    socklog.c, tryto.c, tryto.dist, uncat.c, uncat.dist, wait.h: fix build
    with gcc15: in C23 f() means f(void), consequently signal handler
    functions now are f(int), alloc_free() and byte_zero() parameters need
    casts, buffer_unixwrite() removes "const" from parameter.
  * package/README: update copyright notice.
  * doc/configuration.html: replace links to daemontools faq with runit faq.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * tryto.c: don't report failure if the final try of prog succeeds (thx
    Andras Korn).
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.

socklog 2.1.1
Fri, 11 Oct 2024 00:00:17 +0000
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * doc/index.html: remove reference to socklog 2.0.x.
  * doc/*.html: remove <small>$Id$</small>.
  * debian/: remove; obsolete.
  * pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c, sgetopt.c,
    sgetopt.h, socklog-check.c, socklog-conf.c, socklog.c, subgetopt.c,
    subgetopt.h, trycpp.c, tryflock.c, trypoll.c, trysgact.c, trysgprm.c,
    tryto.c, tryulong64.c, trywaitp.c, uncat.c, x86cpuid.c: incorporate
    lib changes from runit-2.2.0 and adapt socklog programs to properly
    build with modern toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * socklog.c: replace prot_gid() with setgroups() and prot_uid() with
    setuid().

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.

src/tryto.c
#include <unistd.h>
#include <signal.h>
#include "strerr.h"
#include "pathexec.h"
#include "iopause.h"
#include "taia.h"
#include "wait.h"
#include "sig.h"
#include "coe.h"
#include "ndelay.h"
#include "fd.h"
#include "buffer.h"
#include "error.h"
#include "sgetopt.h"
#include "scan.h"

/* defaults */
#define TIMEOUT 180
#define KTIMEOUT 5
#define TRYMAX  5

#define USAGE " [-vp] [-t seconds] [-k kseconds] [-n tries] prog"
#define WARNING "tryto: warning: "
#define FATAL "tryto: fatal: "

const char *progname;
int selfpipe[2];
int try =0;

void sig_child_handler(void) {
  try++;
  write(selfpipe[1], "", 1);
}

void usage () {
  strerr_die4x(1, "usage: ", progname, USAGE, "\n");
}

int main (int argc, char * const *argv, char * const *envp) {
  int opt;
  struct taia now, deadline;
  iopause_fd x[2];
  int pid;
  int rc =111;
  unsigned long timeout =TIMEOUT;
  unsigned long ktimeout =KTIMEOUT;
  unsigned long trymax =TRYMAX;
  int verbose =0;
  char ch;
  int processor =0;
  unsigned int pgroup =0;
  int cpipe[2];

  progname =*argv;

  while ((opt =getopt(argc,argv,"t:k:n:pPvV")) != opteof) {
    switch(opt) {
    case 'V':
      strerr_warn1("$Id$\n", 0);
    case '?':
      usage();
    case 't':
      scan_ulong(optarg, &timeout);
      if (timeout <= 0) timeout =TIMEOUT;
      break;
    case 'k':
      scan_ulong(optarg, &ktimeout);
      if (ktimeout <= 0) ktimeout =KTIMEOUT;
      break;
    case 'n':
      scan_ulong(optarg, &trymax);
      break;
    case 'p':
      processor =1;
      break;
    case 'P':
      pgroup =1;
      break;
    case 'v':
      verbose =1;
      break;
    }
  }
  argv +=optind;
  if (!*argv) usage();

  /* create selfpipe */
  if (pipe(selfpipe) == -1) {
    strerr_die2sys(111, FATAL, "unable to create selfpipe: ");
  }
  coe(selfpipe[0]);
  coe(selfpipe[1]);
  ndelay_on(selfpipe[0]);
  ndelay_on(selfpipe[1]);

  ndelay_on(0);
  if (processor) ndelay_on(4);

  sig_block(sig_pipe);
  sig_block(sig_child);
  sig_catch(sig_child, sig_child_handler);

  /* set timeout */
  taia_now(&now);
  taia_uint(&deadline, timeout);
  taia_add(&deadline, &now, &deadline);
  timeout =0;

  for (;;) {
    int iopausefds;
    char buffer_x_space[BUFFER_INSIZE];
    buffer buffer_x;

    if (processor) {
      buffer_init(&buffer_x, buffer_unixread, 4, buffer_x_space,
		  sizeof buffer_x_space);
    } else {
      buffer_init(&buffer_x, buffer_unixread, 0, buffer_x_space,
		  sizeof buffer_x_space);
    }

    /* start real processor */
    if (pipe(cpipe) == -1) {
      strerr_die2sys(111, FATAL, "unable to create pipe for child: ");
    }
    while ((pid =fork()) == -1) {
      strerr_warn4(WARNING, "unable to fork for \"", *argv, "\" pausing: ",
		   &strerr_sys);
      sleep(5);
    }
    if (!pid) {
      /* child */

      sig_unblock(sig_pipe);
      sig_unblock(sig_child);
      sig_uncatch(sig_child);

      close(cpipe[1]);
      fd_move(0, cpipe[0]);
      if (processor) {
	fd_move(2, 5);
	close(4);
      }
      if (pgroup) setsid();
      pathexec_run(*argv, argv, envp);
      strerr_die2sys(111, FATAL, "unable to start child: ");
    }
    close(cpipe[0]);

    x[0].fd =selfpipe[0];
    x[0].events =IOPAUSE_READ;
    if (processor) {
      fd_move(2, 5);
      x[1].fd =4;
    } else {
      x[1].fd =0;
    }
    x[1].events =IOPAUSE_READ;
    iopausefds =2;

    /* feed + watch child */
    for (;;) {
      int r;
      int i;
      char *s;

      sig_unblock(sig_child);
      iopause(x, iopausefds, &deadline, &now);
      sig_block(sig_child);
      
      while (read(selfpipe[0], &ch, 1) == 1) {}

      taia_now(&now);
      if ((timeout =taia_less(&deadline, &now))) break;
      if (wait_nohang(&rc) == pid) break;
      rc =111;

      r = buffer_feed(&buffer_x);
      if (r < 0) {
	if ((errno == error_intr) || (errno == error_again)) continue;
      }
      if (r == 0) {
	if (processor && (buffer_x.fd == 4)) {
	  x[1].fd =0;
	  buffer_init(&buffer_x, buffer_unixread, 0, buffer_x_space,
		  sizeof buffer_x_space);
	  continue;
	}
	if (iopausefds == 2) {
	  close(cpipe[1]);
	  iopausefds =1;
	}
	continue;
      }
      s =buffer_peek(&buffer_x);
      i =write(cpipe[1], s, r);
      if (i == -1) strerr_die2sys(111, FATAL, "unable to write to child: ");
      if (i < r)
	strerr_die2x(111, FATAL, "unable to write to child: partial write");

      buffer_seek(&buffer_x, r);
    }
    close(cpipe[1]);

    if (timeout) {
      if (wait_nohang(&rc) == pid) break;
      /* child not finished */
      strerr_warn4(WARNING,
		   "child \"", *argv, "\" timed out. sending TERM...", 0);
      kill(pgroup ? -pid : pid, SIGTERM);

      /* ktimeout sec timeout */
      taia_now(&now);
      taia_uint(&deadline, ktimeout);
      taia_add(&deadline, &now, &deadline);
      ktimeout =0;

      for (;;) {
        sig_unblock(sig_child);
        iopause(x, 1, &deadline, &now);
        sig_block(sig_child);

        while (read(selfpipe[0], &ch, 1) == 1) {}

        if (wait_nohang(&rc) == pid) {
	  strerr_warn2(WARNING, "child terminated.", 0);
	  break;
        }
        rc =111;
        taia_now(&now);
        if ((ktimeout =taia_less(&deadline, &now))) break;
      }
      if (ktimeout) {
        strerr_warn4(WARNING, "child \"", *argv,
                     "\" not terminated. sending KILL...", 0);
        kill(pgroup ? -pid : pid, SIGKILL);
      }
      break;
    }
    if (rc == 0) _exit(0);
    if (verbose) strerr_warn2(WARNING, "child crashed.", 0);
    if (lseek(0, 0, SEEK_SET) != 0)
	if (verbose) strerr_warn2(WARNING,
				  "unable to lseek fd 0: ", &strerr_sys);
    if (try >= trymax) break;
    sleep(1);
  }

  if (processor) {
    for (;;) {
      int r;
      char *s;

      r = buffer_feed(buffer_0);
      if (r < 0) {
	if ((errno == error_intr) || (errno == error_again)) continue;
      }
      if (r == 0) {
	break;
      }
      s =buffer_peek(buffer_0);
      buffer_putflush(buffer_1, s, r);
      buffer_seek(buffer_0, r);
    }
  }
  if (timeout) {
    if (processor) strerr_die2x(0, FATAL, "child timed out, giving up.");
    strerr_die2x(100, FATAL, "child timed out, giving up.");
  }
  if (try >= trymax) {
    if (processor) strerr_die2x(0, FATAL, "child crashed, giving up.");
    strerr_die2x(rc >> 8, FATAL, "child crashed, giving up.");
  }
  _exit(0);
}
